### PR TITLE
refactor(context): refactor target type checking for injections into a function

### PR DESCRIPTION
**Pull request by raymondfeng**
Refactor target type checking for injections to a utility function.

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

**Comment by raymondfeng** [comment]
Duplicate of #2747

File: packages/context/src/__tests__/acceptance/class-level-bindings.acceptance.ts [hunk 1]
```diff
@@ -186,7 +186,7 @@ describe('Context bindings - Injecting dependencies of classes', () => {
 
     ctx.bind(STORE_KEY).toClass(Store);
     expect(() => ctx.getSync<Store>(STORE_KEY)).to.throw(
-      'The type of Store.constructor[0] (String) is not a Getter function',
+      'The type of Store.constructor[0] (String) is not Getter function',
```
Comment (bajtos) [inline packages/context/src/__tests__/acceptance/class-level-bindings.acceptance.ts]: Is there any particular reason for removing the article `a`? English is not my mother tongue, but it seems to me that "a Getter function" is more correct grammatically.

(Same comment applies for the second test changed below.)

File: packages/context/src/inject.ts [hunk 2]
```diff
@@ -368,12 +368,43 @@ export namespace inject {
   };
 }
 
+/**
+ * Assert the target type inspected from TypeScript for injection to be the
+ * expected type
```
Comment (bajtos) [inline packages/context/src/inject.ts]: Please explain what happens when the check does not pass: the function throws a descriptive error.

File: packages/context/src/inject.ts [hunk 3]
```diff
@@ -368,12 +368,43 @@ export namespace inject {
...
+ * @param injection
+ * @param session
+ */
 function resolveAsGetter(
   ctx: Context,
   injection: Readonly<Injection>,
   session: ResolutionSession,
 ) {
-  assertTargetIsGetter(injection);
+  checkTargetType(injection, Function, 'Getter function');
```
Comment (bajtos) [inline packages/context/src/inject.ts]:
```suggestion
  checkTargetType(injection, Function, 'a Getter function');
```

File: packages/context/src/inject.ts [hunk 4]
```diff
@@ -385,29 +416,17 @@ function resolveAsGetter(
...
+ */
 function resolveAsSetter(ctx: Context, injection: Injection) {
-  const targetType = inspectTargetType(injection);
-  const targetName = ResolutionSession.describeInjection(injection)!.targetName;
-  if (targetType && targetType !== Function) {
-    throw new Error(
-      `The type of ${targetName} (${targetType.name}) is not a Setter function`,
-    );
-  }
+  const targetName = checkTargetType(injection, Function, 'Setter function');
```
Comment (bajtos) [inline packages/context/src/inject.ts]:
```suggestion
  const targetName = checkTargetType(injection, Function, 'a Setter function');
```

File: packages/context/src/inject.ts [hunk 5]
```diff
@@ -544,7 +547,7 @@ function resolveAsGetterByFilter(
   injection: Readonly<Injection>,
   session: ResolutionSession,
 ) {
-  assertTargetIsGetter(injection);
+  checkTargetType(injection, Function, 'Getter function');
```
Comment (bajtos) [inline packages/context/src/inject.ts]:
```suggestion
  checkTargetType(injection, Function, 'a Getter function');
```

File: packages/context/src/inject.ts (hunk 2 above)
Comment (bajtos) [reply packages/context/src/inject.ts]: Even better, rename this function to `assertTargetType`, I think that's the usual naming conventions for functions that are throwing an error if a check was not passed.

File: packages/context/src/__tests__/acceptance/class-level-bindings.acceptance.ts (hunk 1 above)
Comment (raymondfeng) [reply packages/context/src/__tests__/acceptance/class-level-bindings.acceptance.ts]: IMO, `a` hints an instance, not the type/class. I removed `a` to be consistent with other types, such as `Array` or `ContextView`.

**Pull request merged**